### ISSUE TITLE
flaky `TestInstantQuerySplittingCorrectness`

### ISSUE THREAD

**Issue by seizethedave** [grafana/mimir#10064]
[CI log](https://github.com/grafana/mimir/actions/runs/12092041667/job/33721270921).

```
--- FAIL: TestInstantQuerySplittingCorrectness (26.68s)
    --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z (0.23s)
        --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram (0.26s)
            --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest (0.26s)
                querysharding_test.go:132: 
                    	Error Trace:	/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:132
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:103
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:110
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/split_by_instant_interval_test.go:550
                    	Error:      	Relative error is too high: 1e-12 (expected)
                    	            	        < 0.0163576750[77](https://github.com/grafana/mimir/actions/runs/12092041667/job/33721270921#step:8:78)691952 (actual)
                    	Test:       	TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest
                    	Messages:   	histogram value at position 0 with timestamp 157[78](https://github.com/grafana/mimir/actions/runs/12092041667/job/33721270921#step:8:79)49400000 for series []
FAIL
FAIL	github.com/grafana/mimir/pkg/frontend/querymiddleware	201.859s
```

**Comment by dimitarvdimitrov** [comment]
this happened again. Failure is in the same place. @krajorama @fionaliao do any recent histogram changes come to mind?

<details><summary>Details</summary>
<p>

```
--- FAIL: TestInstantQuerySplittingCorrectness (27.38s)
    --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z (0.28s)
        --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram (0.26s)
            --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest (0.26s)
                querysharding_test.go:132: 
                    	Error Trace:	/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:132
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:103
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:110
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/split_by_instant_interval_test.go:550
                    	Error:      	Relative error is too high: 1e-12 (expected)
                    	            	        < 0.016357675077691952 (actual)
                    	Test:       	TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest
                    	Messages:   	histogram value at position 0 with timestamp 1577849400000 for series []
FAIL
```

</p>
</details>

**Comment by colega** [comment]
It's failing very frequently.

**Comment by chencs** [comment]
This is making doing 2.15 release PRs pretty painful.

**Comment by krajorama** [comment]
Got pinged in slack, but actually I see a different case failing locally:
```
Error:      	Relative error is too high: 1e-12 (expected)
                    	            	        < 0.13827172919340028 (actual)
                    	Test:       	TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_grouping_'by'_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest
                    	Messages:   	histogram value at position 0 with timestamp 1577849400000 for series [{group_1 2}]
```

**Comment by krajorama** [comment]
Seems like it started to fail after #7219 was merged back in January :(

**Comment by krajorama** [comment]
This is duplicate of #7808 . And the workaround in #7504 still works. Maybe we should reconsider applying the workaround as I'm not sure when we'll get to improving the engine itself where the problem originates.

**Comment by dimitarvdimitrov** [comment]
> And the workaround in [#7504](https://github.com/grafana/mimir/pull/7504) still works.

can we merge that workaround and add a linter rule so that we don't use `*promql.StorageSeries` in tests?

**Comment by zenador** [comment]
Consecutive failures ):

https://github.com/grafana/mimir/actions/runs/12400788936/job/34618764750?pr=10277
```
--- FAIL: TestInstantQuerySplittingCorrectness (27.51s)
    --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z (0.23s)
        --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_grouping_'by'_for_native_histogram (0.30s)
            --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_grouping_'by'_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest (0.30s)
                querysharding_test.go:132: 
                    	Error Trace:	/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:132
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:103
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:110
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/split_by_instant_interval_test.go:550
                    	Error:      	Relative error is too high: 1e-12 (expected)
                    	            	        < 0.13116335669173304 (actual)
                    	Test:       	TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_grouping_'by'_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest
                    	Messages:   	histogram value at position 0 with timestamp 1577849400000 for series [{group_1 2}]
```

https://github.com/grafana/mimir/actions/runs/12400788936/job/34619827656?pr=10277
```
 --- FAIL: TestInstantQuerySplittingCorrectness (27.82s)
    --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z (0.25s)
        --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram (0.28s)
            --- FAIL: TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest (0.28s)
                querysharding_test.go:132: 
                    	Error Trace:	/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:132
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:103
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/querysharding_test.go:110
                    	            				/__w/mimir/mimir/pkg/frontend/querymiddleware/split_by_instant_interval_test.go:550
                    	Error:      	Relative error is too high: 1e-12 (expected)
                    	            	        < 0.016357675077691952 (actual)
                    	Test:       	TestInstantQuerySplittingCorrectness/start=2020-01-01T03:00:00Z/sum(rate)_for_native_histogram/*querymiddleware.PrometheusInstantQueryRequest
                    	Messages:   	histogram value at position 0 with timestamp 1577849400000 for series []
```

**Comment by dimitarvdimitrov** [comment]
this should be fixed by https://github.com/grafana/mimir/pull/7504; let's reopen if it props up again
